### ISSUE TITLE
Pydantic's `validate_by_name` and `validation_alias` not recognized in the presence of `alias`

### ISSUE THREAD

**Issue by CharlesMasson** [microsoft/pyright#10741]
Pyright 1.1.403 does not recognize Pydantic's `validate_by_name` config flag, which is shown by the following valid code:

```py
from pydantic import BaseModel, Field


class Model(BaseModel):
    class Config:
        validate_by_name = True

    my_name: int = Field(alias="my_alias")


# error: Argument missing for parameter "my_alias"
# error: No parameter named "my_name"
Model(my_name=1)
```

Likewise, in the presence of `alias`, `validation_alias`, [which takes precedence](https://docs.pydantic.dev/latest/concepts/fields/#mutable-default-values), is not recognized:

```py
from pydantic import BaseModel, Field


class Model(BaseModel):
    my_name: int = Field(alias="my_alias", validation_alias="my_name")


# error: Argument missing for parameter "my_alias"
# error: No parameter named "my_name"
Model(my_name=1)
```

**Comment by erictraut** [comment]
Pyright doesn't have any special-case knowledge of pydantic or its behaviors. It bases its analysis on the type information provided by each library. In the case of pydantic, it uses the [dataclass_transform](https://typing.python.org/en/latest/spec/dataclasses.html#the-dataclass-transform-decorator) feature of the type system to indicate that it has "dataclass-like semantics". To the extent that it deviates from these standard dataclass semantics, standards-compliant type checkers will not be able to reason about these behaviors. The `Config` mechanism is an example of a non-standard behavior, and so is the `validation_alias` feature.

The typing standard does support `alias`, which was added to the spec to support `pydantic` and `attrs`.  The typing spec could be extended to support additional field descriptor capabilities like `validation_alias`, but that would need to go through [the process described here](https://github.com/python/typing-council#decisions). If you would like to see this supported and would like to champion such a change, you're welcome to do so.

If you want your code to type check with standards-based type checkers, you'll need to avoid using features of pydantic that fall outside of the dataclass standard.

If you prefer to continue using these facilities, you can suppress the errors using `# type: ignore` or `# pyright: ignore` comments.
